### PR TITLE
build[cartesian][next]: Bump dace version from 0.14.4 to 0.15.1

### PR DESCRIPTION
**Pull request by edopao**
## Description

Bumping dace version to 0.15.1 affects both `cartesian` and `next` gt4py:
- cartesian: removed try/except for dace backward compatibility
- next: re-enabled some tests that were broken on dace 0.15
- all: fixed and/or suppressed flake8 errors

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -146,7 +146,7 @@ repos:
   ## version = re.search('mypy==([0-9\.]*)', open("constraints.txt").read())[1] 
   ## print(f"#========= FROM constraints.txt: v{version} =========")
   ##]]]
-  #========= FROM constraints.txt: v1.5.1 =========
+  #========= FROM constraints.txt: v1.7.1 =========
   ##[[[end]]]
   rev: v1.5.1  # MUST match version ^^^^ in constraints.txt (if the mirror is up-to-date)
```
Comment (havogt) [inline .pre-commit-config.yaml:151]: just to not forget later, if mypy mirror is up-to-date, then we should 
```suggestion
  rev: v1.7.1  # MUST match version ^^^^ in constraints.txt (if the mirror is up-to-date)
```

**Comment by edopao** [comment]
The cause of the failures in the cartesian tests seems to be that the generate SDFG is not valid, see the discussion on this issue: https://github.com/spcl/dace/pull/1479#issuecomment-1855409819

It was working with dace v0.14.4, but v0.15.1 is no longer able to deal with it. It seems that v0.15.1 does more analysis and transformations on indirection tasklets but, in order to do this, more stringent conditions are required on array access inside tasklets.

File: src/gt4py/eve/datamodels/core.py [hunk 2]
```diff
@@ -1291,7 +1291,7 @@ def _make_datamodel(  # noqa: C901  # too complex but still readable and documen
 
 @utils.optional_lru_cache(maxsize=None, typed=True)
 def _make_concrete_with_cache(
-    datamodel_cls: Type[GenericDataModelT],
+    datamodel_cls: Type,
```
Comment (havogt) [inline src/gt4py/eve/datamodels/core.py:1294]: what's the problem with the concrete type?

File: src/gt4py/next/otf/workflow.py [hunk 3]
```diff
@@ -82,7 +82,9 @@ def replace(self, **kwargs: Any) -> Self:
         if not dataclasses.is_dataclass(self):
             raise TypeError(f"'{self.__class__}' is not a dataclass.")
         assert not isinstance(self, type)
-        return dataclasses.replace(self, **kwargs)  # type: ignore[misc] # `self` is guaranteed to be a dataclass (is_dataclass) should be a `TypeGuard`?
+        return dataclasses.replace(
+            self, **kwargs
+        )  # `self` is guaranteed to be a dataclass (is_dataclass) should be a `TypeGuard`?
```
Comment (havogt) [inline src/gt4py/next/otf/workflow.py:87]: I think this comment was meant for the type ignore and can be removed.

File: src/gt4py/next/program_processors/runners/dace_iterator/__init__.py [hunk 4]
```diff
@@ -260,10 +260,12 @@ def build_sdfg_from_itir(
 
     # visit ITIR and generate SDFG
     program = preprocess_program(program, offset_provider, lift_mode)
-    # TODO: According to Lex one should build the SDFG first in a general mannor.
-    #       Generalisation to a particular device should happen only at the end.
-    sdfg_genenerator = ItirToSDFG(arg_types, offset_provider, column_axis, on_gpu)
+    sdfg_genenerator = ItirToSDFG(arg_types, offset_provider, column_axis)
     sdfg = sdfg_genenerator.visit(program)
+    if sdfg is None:
+        raise RuntimeError(f"visit failed for program {program.id}")
```
Comment (havogt) [inline src/gt4py/next/program_processors/runners/dace_iterator/__init__.py:266]: Update to the style guide for error messages.

File: src/gt4py/eve/utils.py [hunk 5]
```diff
@@ -1536,7 +1536,7 @@ def reduceby(
     ) -> Dict[K, S]:
         ...
 
-    def reduceby(  # type: ignore[misc] # signatures 2 and 4 are not satified due to inconsistencies with type variables
+    def reduceby(  # signatures 2 and 4 are not satified due to inconsistencies with type variables
```
Comment (havogt) [inline src/gt4py/eve/utils.py:1539]: comment was for type ignore

File: src/gt4py/eve/utils.py [hunk 6]
```diff
@@ -1225,7 +1225,7 @@ def unzip(self) -> XIterable[Tuple[Any, ...]]:
             [('a', 'b', 'c'), (1, 2, 3)]
 
         """
-        return XIterable(zip(*self.iterator))  # type: ignore  # mypy gets confused with *args
+        return XIterable(zip(*self.iterator))  # mypy gets confused with *args
```
Comment (havogt) [inline src/gt4py/eve/utils.py:1228]: remove comment for type ignore

File: src/gt4py/next/common.py [hunk 7]
```diff
@@ -189,11 +189,12 @@ def __and__(self, other: UnitRange) -> UnitRange:
...
-            isinstance(value, core_defs.INTEGRAL_TYPES)
-            and value >= self.start
-            and value < self.stop
-        )
+        # TODO(egparedes): use core_defs.IntegralScalar for `isinstance()` checks (see PEP 604)
+        #   and remove int cast, once the related mypy bug (#16358) gets fixed
+        if isinstance(value, core_defs.INTEGRAL_TYPES):
+            return self.start <= cast(int, value) < self.stop
+        else:
+            return False
```
Comment (havogt) [inline src/gt4py/next/common.py:197]: is this change from a merge conflict or is it intended?

**Review by havogt — CHANGES_REQUESTED** [review]

File: src/gt4py/eve/datamodels/core.py (hunk 2 above)
Comment (edopao) [reply src/gt4py/eve/datamodels/core.py:1294]: I must admit that I do not fully understand but it solves this error:
`src/gt4py/eve/datamodels/core.py:817:9: error: Argument 1 to "_make_concrete_with_cache" has incompatible type "type[GenericDataModelT]"; expected "type[Never]"  [arg-type]`

File: src/gt4py/next/common.py (hunk 7 above)
Comment (edopao) [reply src/gt4py/next/common.py:197]: It is intended. Meant to address this error:
```
src/gt4py/next/common.py:194:17: error: Unsupported operand types for <= ("int" and "object")  [operator]
src/gt4py/next/common.py:194:17: error: Unsupported operand types for <= ("Infinity" and "object")  [operator]
src/gt4py/next/common.py:195:17: error: Unsupported operand types for > ("int" and "object")  [operator]
src/gt4py/next/common.py:195:17: error: Unsupported operand types for > ("Infinity" and "object")  [operator]
```

Comment (havogt) [reply src/gt4py/next/common.py:197]: Ok, now I got it (was confused by the `TODO(egparedes)`)

**Review by havogt — APPROVED** [review]
I'd change the typing fix, see comment. Otherwise looks good.

**Pull request merged**